### ISSUE TITLE
torch.unique undocumented behaviour

### ISSUE THREAD

**Issue by zurkin1** [pytorch/pytorch#150124]
### 📚 The doc issue

The following page is wrong and gives unclear explanation for torch.unique (which affects torch.unique_consecutive as well). Using the example from this page:

a = torch.tensor([[[1, 1, 0, 0],[1, 1, 0, 0],[0, 0, 1, 1],],[[0, 0, 1, 1],[0, 0, 1, 1],[1, 1, 1, 1],],[[1, 1, 0, 0],[1, 1, 0, 0],[0, 0, 1, 1],],])

torch.unique(a, dim=0):
- Since the two matrices (a[0, :, :] == a[2, :, :]).all() we remove one and get:
tensor([[[0, 0, 1, 1],[0, 0, 1, 1],[1, 1, 1, 1]],[[1, 1, 0, 0],[1, 1, 0, 0],[0, 0, 1, 1]]])
This part of the documentation is correct.

torch.unique(a, dim=1):
Here each of the tensors is treated as one of the elements to apply the unique operation upon.
tensor([[[0, 0, 1, 1],
          [1, 1, 0, 0]],
         [[1, 1, 1, 1],
          [0, 0, 1, 1]],
         [[0, 0, 1, 1],
          [1, 1, 0, 0]]])
The results is removed duplicate rows in each matrix (separately) in contrast to the documentation (that talks about some minor matrix comparison a[:, idx, :], but it is wrong).

torch.unique(a, dim=2):
Same as before. Comparing complete columns within each matrix. Minor matrices are irrelevant.

### Suggest a potential alternative/fix

Suggestion:
1) Please correct the documentation.
2) Many times we care about unique (or unique_consecutive) of *elements* (not complete rows or columns). In the case of dim=1 we want unique elements on every row of the matrix (each matrix processed separately). This feature is currently not supported and pretty tricky to program. Below is my solution (workaround) for that, but it would be best to add as a parameter to unique, for example:
torch.unique(input, sorted=True,..., dim1, dim2)

And we always require the cardinality of |dim2| < |dim1|
So if we operate on dim1==matrix level (for example), we can check uniqueness of rows, columns, or elements.
In the same way if dim1==1 (for example rows of a matrix) we can only check uniqueness of elements.
This provides the maximal flexibility for the user.

--------
My implementation for element uniqueness accross columns of a batch of tensors (in my case binary tensors):
def process_vertical(bin_mats):
    """
    Processes vertical lines of binary recurrence matrices for a batch.
    Args:
        bin_mats: A batch of binary recurrence matrices (shape: [batch_size, n_vectors, n_vectors]).
    Returns:
        counts: A vector of counts of 1's per column, seperated by a -1 marker.
    """
    # Add a dummy top row of -1's to split sequence of 1's accross columns and mark the end of the column.
    # 0, 0: No padding on columns (left and right)
    # 0: No padding at the top
    # 1: Add one row at the bottom
    padded_mats = torch.nn.functional.pad(bin_mats, (0, 0, 0, 1), value=-1)

    # Transpose the matrices for column-wise processing
    padded_mats = padded_mats.transpose(1, 2)  # Shape: [batch_size, n_vectors, n_vectors]
    unique_vec, counts_vec = torch.unique_consecutive(padded_mats, return_counts=True)
    
    # Filter counts to include only counts of 1's and -1's.
    counts_of_ones = counts_vec * unique_vec
    
    return bin_mats.shape, counts_of_ones

**Comment by albanD** [comment]
> The results is removed duplicate rows in each matrix (separately) in contrast to the documentation (that talks about some minor matrix comparison a[:, idx, :], but it is wrong).

Could you give more details on that? From re-reading the doc it does sound correct to me?

> Many times we care about unique (or unique_consecutive) of elements (not complete rows or columns)

This is very problematic to do because each row (or column) might have a different number of unique elements and so the output is not a rectangular Tensor like we usually have. So this would be a tricky function to add to core.

**Comment by zurkin1** [comment]
Please look at the documentation for torch.unique, at the example section for the case torch.unique(a, dim=1).

>>> # For `torch.unique(a, dim=1)`, each of the tensors `a[:, idx, :]` are
>>> # compared. `a[:, 0, :]` and `a[:, 1, :]` match each other, so one of
>>> # them will be removed.

This is wrong! it's not how torch.unique(a, dim=1) works and the result is not relevant to a[:,i,:] (3 x 4 matrices). The result is of shape torch.Size([3, 2, 4]) and the effective operation is:
- For each matrix:
  - Compare and remove duplicate rows.

The same for the dim=1 documentation.
---------------------
Suggested fix (adding another dim parameter for comparison "scale"):
"This is very problematic to do because each row (or column) might have a different number of unique elements and so the output is not a rectangular Tensor like we usually have. So this would be a tricky function to add to core."
You can always pad with zero's (or a configurable integer, this is what I did). Comparing elements is much more needed IMHO than comparing complete rows or complete matrices.

**Comment by albanD** [comment]
I'm sorry I don't follow you explanation. The example does match the actual behavior of the function and the description there describes what the function is doing accurately as well.

**Comment by zurkin1** [comment]
Hi @albanD,

I think we're having some miscommunication about what torch.unique(a, dim=1) actually does versus what the documentation says it does. Let me try to explain this more clearly with a concrete example.

Let's use the same tensor from the documentation:

```python
import torch

a = torch.tensor([[[1, 1, 0, 0],[1, 1, 0, 0],[0, 0, 1, 1]],     # First 2D matrix
                 [[0, 0, 1, 1],[0, 0, 1, 1],[1, 1, 1, 1]],               # Second 2D matrix
                 [[1, 1, 0, 0],[1, 1, 0, 0],[0, 0, 1, 1]]])             # Third 2D matrix

print("Tensor shape:", a.shape)  # Should be torch.Size([3, 3, 4])

The documentation currently says for torch.unique(a, dim=1):
"For torch.unique(a, dim=1), each of the tensors a[:, idx, :] are compared. a[:, 0, :] and a[:, 1, :] match each other, so one of them will be removed."
But when we actually run:

```python
result = torch.unique(a, dim=1)
print("Result shape:", result.shape)
print(result)

we get:
```python
Result shape: torch.Size([3, 2, 4])
b = tensor([[[0, 0, 1, 1],
         [1, 1, 0, 0]],

        [[1, 1, 1, 1],
         [0, 0, 1, 1]],

        [[0, 0, 1, 1],
         [1, 1, 0, 0]]])

What's happening here is NOT that it's comparing entire a[:, idx, :] matrices across batches as the documentation suggests. 
(Note that, for example, a[:,0, :] is a matrix of shape (3x4) that simply takes the first row from each one of the original 2D matrices in a. It has nothing to do with the resulting matrix b above).

The explanation instead:
It's treating each batch index (original 2D matrix) separately
Within each 2D matrix, it's removing duplicate rows, and thats it. We get 3 "smaller" 2D matrices.
For example, in the first 3×4 matrix (a[0]), rows 0 and 1 are identical, so one is removed

To clarify with a visualization:
Input tensor a[0]:
'''python
[[1, 1, 0, 0],  # Row 0
 [1, 1, 0, 0],  # Row 1 (duplicate of row 0)
 [0, 0, 1, 1]]  # Row 2

After torch.unique(dim=1) on a[0] (the matrix above):
```python
[[1, 1, 0, 0],  # Kept row 0
 [0, 0, 1, 1]]  # Kept row 2

I believe the documentation should be updated to clarify that torch.unique(a, dim=1) removes duplicate rows within each batch, rather than comparing entire a[:, idx, :] slices as it currently states.
Would it help if I prepared a more detailed explanation with more examples? I'm happy to contribute to improving the documentation if that would be helpful.

**Comment by albanD** [comment]
What if you change the example to
```
a = torch.tensor([[[1, 1, 0, 0],[1, 1, 1, 0],[0, 0, 1, 1]],     # First 2D matrix
                 [[0, 0, 1, 1],[0, 0, 1, 1],[1, 1, 1, 1]],               # Second 2D matrix
                 [[1, 1, 0, 0],[1, 1, 1, 0],[0, 0, 1, 1]]])             # Third 2D matrix
```
such that the row are still duplicate but a[:, 0, :] != a[:, 1, :] are not equal anymore.

**Comment by zurkin1** [comment]
Again, I don't think a[:,i,:] are relevant at all. In your example:
In [23]: b[:,0,:]
Out[23]:
tensor([[1, 1, 0, 0],
           [0, 0, 1, 1],
           [1, 1, 0, 0]])

In [24]: b[:,1,:]
Out[24]:
tensor([[1, 1, 1, 0],
           [0, 0, 1, 1],
           [1, 1, 1, 0]])

In [25]: b[:,2,:]
Out[25]:
tensor([[0, 0, 1, 1],
           [1, 1, 1, 1],
           [0, 0, 1, 1]])

but
In [27]: torch.unique(b, dim=1)
Out[27]:
tensor([[[0, 0, 1, 1],
             [1, 1, 0, 0],
             [1, 1, 1, 0]],

        [[1, 1, 1, 1],
         [0, 0, 1, 1],
         [0, 0, 1, 1]],

        [[0, 0, 1, 1],
         [1, 1, 0, 0],
         [1, 1, 1, 0]]])

Doesn't seem to have anything with b[:,i,:] values.
